candidatepair.go

```diff
@@ -141,8 +156,101 @@ func (p *CandidatePair) TotalRoundTripTime() float64 {
+// https://www.w3.org/TR/webrtc-stats/#dom-rtcicecandidatepairstats-requestsreceived
+func (p *CandidatePair) RequestsReceived() uint64 {
+	return atomic.LoadUint64(&p.requestsReceived)
+}
+
+// RequestsSent returns the total number of connectivity checks sent
+// https://www.w3.org/TR/webrtc-stats/#dom-rtcicecandidatepairstats-requestssent
+func (p *CandidatePair) RequestsSent() uint64 {
+	return atomic.LoadUint64(&p.requestsSent)
+}
+
 // ResponsesReceived returns the total number of connectivity responses received
 // https://www.w3.org/TR/webrtc-stats/#dom-rtcicecandidatepairstats-responsesreceived
 func (p *CandidatePair) ResponsesReceived() uint64 {
 	return atomic.LoadUint64(&p.responsesReceived)
 }
+
+// ResponsesSent returns the total number of connectivity responses sent
+// https://www.w3.org/TR/webrtc-stats/#dom-rtcicecandidatepairstats-responsessent
+func (p *CandidatePair) ResponsesSent() uint64 {
+	return atomic.LoadUint64(&p.responsesSent)
+}
+
+// FirstRequestSentAt returns the timestamp of the first connectivity check sent.
+func (p *CandidatePair) FirstRequestSentAt() time.Time {
+	if v, ok := p.firstRequestSentAt.Load().(time.Time); ok {
+		return v
+	}
+
+	return time.Time{}
+}
+
+// LastRequestSentAt returns the timestamp of the last connectivity check sent.
+func (p *CandidatePair) LastRequestSentAt() time.Time {
+	if v, ok := p.lastRequestSentAt.Load().(time.Time); ok {
+		return v
+	}
+
+	return time.Time{}
+}
+
+// FirstReponseReceivedAt returns the timestamp of the first connectivity response received.
+func (p *CandidatePair) FirstReponseReceivedAt() time.Time {
+	if v, ok := p.firstReponseReceivedAt.Load().(time.Time); ok {
+		return v
+	}
+
+	return time.Time{}
+}
+
+// LastResponseReceivedAt returns the timestamp of the last connectivity response received.
+func (p *CandidatePair) LastResponseReceivedAt() time.Time {
+	if v, ok := p.lastResponseReceivedAt.Load().(time.Time); ok {
+		return v
+	}
+
+	return time.Time{}
+}
+
+// FirstRequestReceivedAt returns the timestamp of the first connectivity check received.
+func (p *CandidatePair) FirstRequestReceivedAt() time.Time {
+	if v, ok := p.firstRequestReceivedAt.Load().(time.Time); ok {
+		return v
+	}
+
+	return time.Time{}
+}
+
+// LastRequestReceivedAt returns the timestamp of the last connectivity check received.
+func (p *CandidatePair) LastRequestReceivedAt() time.Time {
+	if v, ok := p.lastRequestReceivedAt.Load().(time.Time); ok {
+		return v
+	}
+
+	return time.Time{}
+}
+
+// UpdateRequestSent increments the number of requests sent and updates the timestamp.
+func (p *CandidatePair) UpdateRequestSent() {
+	atomic.AddUint64(&p.requestsSent, 1)
+	now := time.Now()
+	p.firstRequestSentAt.CompareAndSwap(nil, now)
+	p.lastRequestSentAt.Store(now)
+}
+
+// UpdateResponseSent increments the number of responses sent.
+func (p *CandidatePair) UpdateResponseSent() {
+	atomic.AddUint64(&p.responsesSent, 1)
```

Comment:
Is there any value to adding first/last response sent time?

With this PR, we have
- first/last time of request sent
- first/last time of request received
- first/last time of response received

wondering if there is more to learn by recording first/last time of response sent?

Reply:
I had it but deleted at last since the agent always send response when request is received, so the timpstamp is useless (the `responseSent` has same issue too but it is in the specification https://www.w3.org/TR/webrtc-stats/#dom-rtcicecandidatepairstats-responsessent so I keep it).

Reply:
got it, thank you @cnderrauber